File: 373.查找和最小的k对数字/kSmallestPairs.go
package problem0373

func kSmallestPairs(nums1 []int, nums2 []int, k int) [][]int {
	heap := make([][]int, 0)
	for _, a := range nums1 {
		for _, b := range nums2 {
			if len(heap) < k {
				push(&heap, []int{a, b})
			} else if less(heap[0], []int{a, b}) {
				pop(&heap)
				push(&heap, []int{a, b})
			}
		}
	}
	return heap
}

func push(arr *[][]int, x []int) {
	*arr = append(*arr, x)
	siftUp(*arr, len(*arr)-1)
}

func pop(arr *[][]int) {
	lastIndex := len(*arr) - 1
	swap(*arr, 0, lastIndex)
	siftDown(*arr, 0, lastIndex)
	*arr = (*arr)[0:lastIndex]
}

func siftDown(arr [][]int, i int, length int) {
	leftIndex := i*2 + 1
	if leftIndex >= length {
		return
	}
	if leftIndex+1 < length && less(arr[leftIndex+1], arr[leftIndex]) {
		leftIndex++
	}
	if less(arr[leftIndex], arr[i]) {
		swap(arr, leftIndex, i)
		siftDown(arr, leftIndex, length)
	}
}

func siftUp(arr [][]int, i int) {
	pIndex := (i - 1) / 2
	if less(arr[i], arr[pIndex]) {
		swap(arr, pIndex, i)
		siftUp(arr, pIndex)
	}
}

func less(a []int, b []int) bool {
	return a[0]+a[1] > b[0]+b[1]
}

func swap(arr [][]int, i int, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

File: 373.查找和最小的k对数字/kSmallestPairs.js
// 典型送分题
function kSmallestPairs(nums1, nums2, k) {
  const heap = new Heap([], (a, b) => {
    return a[0] + a[1] < b[0] + b[1]
  })
  for (const num1 of nums1) {
    for (const num2 of nums2) {
      heap.push([num1, num2])
    }
  }
  const result = []
  for (let i = 0; i < k; i++) {
    result.push(heap.pop())
  }
  return result.filter(Boolean)
}

// 堆的实现
class Heap {
  constructor(data = [], less) {
    this.data = data
    this.less = less || ((a, b) => a < b)
    // 对初始数组进行堆化操作：从最后一个父节点开始，依次下沉
    if (this.length) {
      for (let p = (this.length - 2) >> 1; p >= 0; p--) {
        this._down(p)
      }
    }
  }
  get length() {
    return this.data.length
  }
  peak() {
    return this.data[0]
  }
  push(val) {
    this.data.push(val)
    this._up(this.length - 1)
  }
  pop() {
    if (!this.length) {
      return
    }
    this._swap(0, this.length - 1)
    const popItem = this.data.pop()
    this._down(0)
    return popItem
  }
  // 删除指定值的操作，通常堆是不含这个操作的，针对特定题目，添加上这个操作
  remove(val) {
    let index = -1
    for (let i = 0; i < this.length; i++) {
      if (this.data[i] === val) {
        index = i
        break
      }
    }
    if (index === -1) return false
    this._swap(index, this.length - 1)
    this.data.pop()
    this._up(index)
    this._down(index)
    return true
  }
  _swap(i, j) {
    [this.data[i], this.data[j]] = [this.data[j], this.data[i]]
  }
  // 递归上浮
  _up(i) {
    if (i <= 0) return
    const pIndex = (i - 1) >> 1
    if (this.less(this.data[i], this.data[pIndex])) {
      this._swap(i, pIndex)
      this._up(pIndex)
    }
  }
  // 递归下沉
  _down(i) {
    let leftIndex = i * 2 + 1
    if (leftIndex >= this.length) {
      return
    }
    if (leftIndex + 1 < this.length && this.less(this.data[leftIndex+1], this.data[leftIndex])) {
      leftIndex++
    }
    if (this.less(this.data[leftIndex], this.data[i])) {
      this._swap(leftIndex, i)
      this._down(leftIndex)
    }
  }
}
